tokio/src/task/builder.rs

```diff
@@ -0,0 +1,95 @@
+#![allow(unreachable_pub)]
+use crate::task::JoinHandle;
+use std::future::Future;
+
+/// Factory which is used to configure the properties of a new task.
+///
+/// Methods can be chained in order to configure it.
+///
+/// Currently, there is only one configuration option:
+///
+/// - [`name`], which specifies an associated name for
+///   the task
+///
+/// There are three types of task that can be spawned from a Builder:
+/// - [`spawn_local`] for executing futures on the current thread
+/// - [`spawn`] for executing [`Send`] futures on the runtime
+/// - [`spawn_blocking`] for executing blocking code in the
+///   blocking thread pool.
+#[derive(Default, Debug)]
+pub struct Builder<'a> {
+    name: Option<&'a str>,
+}
+
+impl Builder<'_> {
+    /// Creates a new task builder
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Assigns a name to the task which will be spawned
+    pub fn name<'b>(&self, name: &'b str) -> Builder<'b> {
+        Builder { name: Some(name) }
+    }
+
+    /// Spawns a task on the executor.
+    ///
+    /// See
+    /// [`task::spawn`][crate::task::spawn] for more details
+    #[cfg_attr(tokio_track_caller, track_caller)]
+    pub fn spawn<T>(self, future: T) -> JoinHandle<T::Output>
+    where
+        T: Future + Send + 'static,
+        T::Output: Send + 'static,
+    {
+        super::spawn::spawn_inner(future, self.name)
+    }
+
+    /// Spawns a task on the current thread.
+    ///
+    /// See [`task::spawn_local`][crate::task::spawn_local]
+    /// for more details
+    #[cfg_attr(tokio_track_caller, track_caller)]
```

Comment:
For the caller location to be tracked, do we need to call `Location::caller()` in the method proper or is the location passed to inner calls?

Reply:
My read of (https://doc.rust-lang.org/reference/attributes/codegen.html#behavior)[this] is that it just makes the annotated function "transparent" to the walked Location, so inner calls don't report this function, but its caller. I'm new to this interface though — did I misunderstand it?

Reply:
Not sure what you mean exactly, but here: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=9fe91c5452a8add4f3b9a45e688ec3c7

Without `#[track_caller]` on `fn two()`, `Location::caller()` isn't what we want. If `fn two()` does have `#[track_caller]` then it is passed through `one()` and reports the line within `fn main()`.

Reply:
So it looks like you are correct.